src/server/pkg/storage/chunk/writer.go

```diff
@@ -403,70 +468,138 @@ func (w *Writer) ChunkCount() int64 {
 func (w *Writer) Write(data []byte) (int, error) {
+	a := w.annotations[len(w.annotations)-1]
 	var written int
-	for w.buf.Len()+len(data) >= bufSize {
-		i := bufSize - w.buf.Len()
-		w.buf.Write(data[:i])
-		w.writeByteSet()
+	for w.bufSize+len(data) >= bufSize {
+		i := bufSize - w.bufSize
+		a.buf.Write(data[:i])
+		w.stats.taggedBytesSize += int64(i)
+		w.writeDataSet()
+		a = w.annotations[len(w.annotations)-1]
 		written += i
 		data = data[i:]
 	}
-	w.buf.Write(data)
+	a.buf.Write(data)
+	w.bufSize += len(data)
+	w.stats.taggedBytesSize += int64(len(data))
 	written += len(data)
-	w.stats.annotatedBytesSize += int64(written)
 	return written, nil
 }
 
-func (w *Writer) writeByteSet() {
+func (w *Writer) writeDataSet() {
+	w.finishTag()
 	prev := w.prev
 	next := &chanSet{
-		bytes: make(chan *byteSet, 1),
-		done:  make(chan struct{}),
-	}
-	byteSet := &byteSet{
-		data:        w.buf.Bytes(),
-		annotations: w.annotations,
+		dataSet: make(chan *dataSet, 1),
+		done:    make(chan struct{}),
 	}
+	dataSet := &dataSet{annotations: w.annotations}
 	w.eg.Go(func() error {
-		return w.newWorkerFunc(w.cancelCtx, newPrevChanSet(prev), newNextChanSet(next)).run(byteSet)
+		return w.newWorkerFunc(w.cancelCtx, newPrevChanSet(prev), newNextChanSet(next)).run(dataSet)
 	})
 	w.prev = next
-	w.buf = &bytes.Buffer{}
-	w.annotations = splitAnnotations(w.annotations)
+	lastA := w.annotations[len(w.annotations)-1]
+	_, a := splitAnnotation(lastA, lastA.buf.Len())
+	w.annotations = []*Annotation{a}
+	w.bufSize = 0
 }
 
-// Copy does a cheap copy from a reader to a writer.
-func (w *Writer) Copy(r *Reader, n ...int64) error {
-	c, err := r.ReadCopy(n...)
-	if err != nil {
-		return err
+// Copy copies data from a data reader to the writer.
+// The copy will either be by reading the referenced data, or just
+// copying the data reference (cheap copy).
+func (w *Writer) Copy(dr *DataReader) error {
+	lastA := w.annotations[len(w.annotations)-1]
+	lastA.drs = append(lastA.drs, dr)
+	w.bufSize += int(dr.Len())
+	if w.bufSize > bufSize {
+		w.writeDataSet()
 	}
-	return w.WriteCopy(c)
+	return nil
 }
 
-// WriteCopy writes copy data to the writer.
-func (w *Writer) WriteCopy(c *Copy) error {
-	if _, err := io.Copy(w, c.before); err != nil {
-		return err
+// Close closes the writer.
+func (w *Writer) Close() error {
+	// Write out the last data set.
+	if w.bufSize > 0 {
+		w.writeDataSet()
 	}
-	for _, chunkRef := range c.chunkRefs {
-		w.stats.chunkCount++
-		// (bryce) might want to double check if this is correct.
-		w.stats.annotatedBytesSize += chunkRef.SizeBytes
-		updateAnnotations(chunkRef, nil, w.annotations)
-		if err := w.f(chunkRef, w.annotations); err != nil {
-			return err
+	// Signal to the last worker that it is last.
+	if w.prev != nil {
+		close(w.prev.dataSet)
+		w.prev = nil
+	}
+	// Wait for the workers to finish.
+	return w.eg.Wait()
+}
+
+func joinAnnotations(as []*Annotation, a *Annotation) []*Annotation {
```

Comment:
Can we call this `appendAnnotations`? I think that's a more descriptive name, that more closely matches what's going on, I think of joins as actually considering the content that's being joined rather than just, well appending them together.

Reply:
Similarly could this be made into a variadic function like the builtin `append`?

Reply:
It merges the new annotation with the last one when they have the same `Meta`. So, it is considering the content. If you interpret it as just appending to a slice, then that would be incorrect. I could expose a variadic variant of this, but I cannot think of a use case for it now or in the future. We will always be operating on one annotation at a time.

Reply:
I see, that makes sense, I think `join` is actually a fine name.`